src/App.js

```diff
@@ -1,35 +1,73 @@
-import React, { useState } from "react";
+import React, { useState, useEffect } from "react";
 import TodoList from "./TodoList";
 import AddTodoForm from "./AddTodoForm.js";
 
 //custom hook
-const useSemiPersistentState = (initialState, key) => {
-  // Default state to read "savedTodoList" from localStorage
-  const [state, setState] = useState(
-    JSON.parse(localStorage.getItem(key)) || initialState
-  );
-  
-  React.useEffect(() => {
-    // Save todoList to localStorage with the key "savedTodoList"
-    localStorage.setItem(key, JSON.stringify(state));
-    }, [state, key]);
+// const useSemiPersistentState = (initialState, key) => {
+//   // Default state to read "savedTodoList" from localStorage
+//   const [state, setState] = useState(
+//     JSON.parse(localStorage.getItem(key)) || initialState
+//   );
 
-    return [state, setState]
-}
+//   React.useEffect(() => {
+//     // Save todoList to localStorage with the key "savedTodoList"
+//     localStorage.setItem(key, JSON.stringify(state));
+//   }, [state, key]);
+
+//   return [state, setState];
+// };
 
 function App() {
-  const [todoList, setTodoList] = useSemiPersistentState([], "savedTodoList");
+  // Default state to read "savedTodoList" from localStorage
+  const initialState = JSON.parse(localStorage.getItem("savedTodoList")) || [];
+  const [todoList, setTodoList] = useState(initialState);
+  const [isLoading, setIsLoading] = useState(true);
+
+  useEffect(() => {
+    const fetchData = new Promise((resolve, reject) => {
+      // Simulate loading delay with a timeout
+      setTimeout(() => {
+        resolve({
+          data: {
+            todoList: initialState, // Set the initial/default list state
+          },
+        });
+      }, 2000); // 2 seconds delay
+    });
+
+    fetchData.then((result) => {
+      setTodoList(result.data.todoList);
```

Comment:
This works fine for now. However, this can cause a bug. Since you are not passing a copy to your setter function a render might not be triggered. To correctly update state and trigger a render pass a copy of the object, in this case array. eg `setTodoList([...result.data.todoList])